Add data binding interfaces

dpgil opened the pull request:
This change adds interfaces which facilitate data binding between different components. This change adds the following interfaces:
- IDataProvider: to be implemented by a component that wants to expose some data to other components
- IDataObserver: to be implemented by a component that wants to listen to data changes in other components
- IDataResult: the contract defining the type passed through data binding between components, provides IData along with additional metadata, such as hint text, friendly name, etc.
- IData: the most primitive data type in data binding, i.e. the computed value from a data binding
- ICustomData, ICollection, ITable: our start of data shapes, allowing more complex types from a data binding relationship

This is going to break the tests in the example package. We will need to update them to conform to the new contracts

dpgil commented:
@DLehenbauer I updated getValue to use Pending, and changed getProperties to take a string and to be async.

Do we want getProperties to use Pending as well?
Also, let me know if you think some of the naming should change.

jack-williams commented on packages/core/nano/src/types.ts:
```diff
@@ -1,22 +1,148 @@
+  /**
+   * Return the raw value of this object.
+   */
+  getValue: () => Pending<IData | undefined>;
```
I think this should be `IData | Pending<IData>` (and you can throw in the `undefined` if you need it). The `undefined` was there in the `now` function because that was a way of querying the immediate value of a component without issuing a data binding, so if it wasn't ready yet we couldn't return a pending because we have no consumer to send it to.

I think in these cases an `IDataResult` implies a databinding so it's analogues to the request function, so it will just return pending and then later an update to the consumer.

DLehenbauer commented:
FYI - Here's the process I recommend:
1.  We each take turns iterating on 'dpgil:patch-1' to merge ideas.
2.  Anything that doesn't make the cut for Tiny-Calc goes to prague/framework-definitions (possibly as extensions of the tiny-calc interfaces.  e.g., IDataProvider could extend IProducer)
3.  @jack-williams has the final say.

DLehenbauer commented on packages/core/nano/src/types.ts:
```diff
@@ -1,22 +1,139 @@
+  /**
+   * Invoked when this data provider is deleted.
+   */
+  onDeleted(): void;
```
What is the purpose of the 'onDeleted()' callback?

In other systems I've work on, a producer releases resources when the number of consumers drops to zero.  Would that work for your scenario?

dpgil replied:
How are consumers notified when a producer is deleted? In a more practical example, say we have a Formula bound to the name of a table. When the table is deleted, we don't want the formula to show the table name anymore, we probably want it to show some broken reference UI. Also, we'd want to make sure the producer is unregistered from the DataBindingService (not a service we're adding into this repo). Can those concepts still be addressed without onDeleted?

DLehenbauer commented on packages/core/nano/src/types.ts:
```diff
@@ -1,22 +1,139 @@
+  /**
+   * Invoked whenever the provider this is bound to is deleted.
+   */
+  onProducerDeleted: () => void;
```
What is the purpose of the onProducerDeleted callback?  In other systems I've worked on, I had a similar notification that told a consumer when a producer became fixed (i.e., wouldn't change again), giving the consumer an opportunity to drop any tracking related state.

dpgil replied:
How are consumers notified when a producer is deleted? In a more practical example, say we have a Formula bound to the name of a table. When the table is deleted, we don't want the formula to show the table name anymore, we probably want it to show some broken reference UI. Also, we'd want to make sure the producer is unregistered from the DataBindingService (not a service we're adding into this repo). Can those concepts still be addressed without onDeleted?

jack-williams replied:
I think the implicit assumption here is that all producers that could be deleted are actually some property nested in another producer, and that a deletion is just a notification from the outer producer. Something like:
```ts
type FormulaRootContext = IProducer< { TableName: IProducer<ITable> | MissingRef } >;
```
The formula would be subscribed to the root context producer and when table `TableName` is deleted it gets notified that the property has changed its value from a table to 'missing', for some representation of missing.

DLehenbauer replied:
One idea is to use the revoker pattern:
```ts
class RevokableProducer<T> implements IProducer<T>, IConsumer<T> {
    private consumers = new Set<IConsumer<T>>();

    constructor(private readonly producer: IProducer<T>) {}

    private get isRevoked() { return !this.consumers; }

    public read(consumer, ...) {
        if (this.isRevoked) { return new Error(`Not available.`); }
        this.consumers.add(consumer);

        return this.producer.read(this, ..);
    }

    public removeConsumer(consumer) {
        this.consumers.delete(consumer);
        if (this.consumers.count === 0) { this.producer.unsubscribe(this); }
    }

    public notify(...) {
        assert(!this.revoked);
        for (const consumer of consumers) { consumer.notify(...); }
    }

    public revoke() {
        this.revoked = true;
        this.notify(this, '*', new Error('Not available.'));      // Invalidate all properties from this producer
        this.consumers = undefined;
        this.producer.unsubscribe(this);
    }   
}
```

dpgil replied:
I see, that makes sense. So we just treat deletion as another property change, albeit a catastrophic one?

DLehenbauer replied:
Right.  You can also probably get more mileage out of the wrapper object (e.g., a general adapter for IProducers that knows how to create IDataResults for your app's internal use.)

DLehenbauer commented on packages/core/nano/src/types.ts:
```diff
@@ -1,22 +1,139 @@
+ * It provides query functions for retrieving values, hints, friendly names, and provides
+ * basic traverse APIs to move to previous or next level results.
+ */
+export interface IDataResult {
```
I'm not yet sure how to reconcile IDataResult.

I'm currently thinking that the Intellisense related APIs should be a separate interface at the framework level that an IProducer optionally implements, like:

```ts
class Producer implements IProducer<T>, IIntellisenseProvider<T> {
    getProperties(query: string) => Promise<string[]>;
    getHint(property: string): string;
    getFriendlyName(property: string): string;

    // ..plus regular IProducer stuff
}
```
...unless Jack has any use for these APIs?

jack-williams replied:
If my understanding is correct `IDataResult` is a static value and does not contain any observables/producers within it - in that sense `IDataResult` is really just an app specific value that a `IProducer` may provide. For the `IProducer` type I think there are two constraints we want to juggle (there may be more). First, a conceptually simple producer should be simple to implement, so we don't want to make the base interfaces too heavy. Second, I think there should be some standard way to explore a network of producers.

I'm inclined to say that `IProducer` should have the method `getProperties(query: string) => Promise<string[]>;` with the contract that the array of strings is a lower bound on the set of subscribable properties (perhaps even make the method optional). This way someone could quickly implement something like `getProperties() => Promise.resolve([])` without violating the contract.

I think the other things would belong in supplementary interfaces as @DLehenbauer suggests.

jack-williams replied:
And I don't have any uses for these API's atm. Those might likely one I would need is a way to enumerate properties and possibly get their types, but that is further down the line.

DLehenbauer commented on packages/core/nano/src/types.ts:
```diff
@@ -1,22 +1,139 @@
+  /**
+   * Return The dataBindingReference for this instance of Producer.
+   */
+  getBindingReference: () => any;
```
I think getBindingReference is just the key used to read the value?

dpgil replied:
Sort of. Currently we have the Consumer hold on to the IDataResult that it's bound to. Since we can't store it over the wire, the binding reference is a serializable object that the Producer knows know how to interpret to find the right IDataResult.

I think it makes more sense though for the Producer to keep track of what values the Consumers are bound to. We had been planning to redesign our use of data binding to get rid of this function so I'm all for removing it

DLehenbauer commented on packages/core/nano/src/types.ts:
```diff
@@ -1,22 +1,139 @@
+   * @returns an instance of IDataResult that represent the previous level (parent) of this object,
+   * or undefined if this is the top level object.
+   */
+  getParent: () => IDataResult | undefined;
```
I don't follow 'getParent()'.  Is this related to the parser state at point we evaluated this result?  (Otherwise, wouldn't it be getParents?)

dpgil replied:
Yeah, it's so we can reconstruct the full formula given one piece of it. I'm not sure if this should be more of an implementation detail for the Formula component specifically.

DLehenbauer commented:
(Yielding my turn to @jack-williams.)

jack-williams commented on packages/core/nano/src/types.ts:
```diff
@@ -1,22 +1,139 @@
+   * @param dataBindingReference - An instance of dataBindingReference, a reference
+   * to some data in the bound component.
+   */
+  read<K extends keyof T>(consumer: IConsumer<T>, property: K, ...args: any[]): T[K] | Pending<T[K]>;
```
Agree

jack-williams commented:
Just some context on the producer type I had. A scenario I was imagining was something like a complex object (like a file) modelled as a producer, so the domain interface might be something like:
```ts
interface File {
  size: number;
  name: string;
  path: string;
}
```
I strayed away from having `Observable<File>` because a consumer that was only interested in low update rate properties such as `name` would be penalised by frequently updating things like `size`. One solution is to push the change down into the leaves:
```ts
interface File {
  size: Observable<number>;
  name: Observable<string>;
  path: Observable<string>;
}
```
But the concern I had with this approach was that 1) you need to allocate many observer functions for each property, 2) you lose the ability to enforce simultaneous update. If `name` and `path` change at the same timestep it's possible to get into an impossible state between the receipt of first update of `name`, and `before` the incoming update of `path`. You can use observable combinators but it's an extra level of indirection.

My intent for `Producer` was to be viewed as a reactive map, or event source where each property corresponded to an event type. Doing a `read` on property `K` would subscribe the consumer to that property on the producer, and the purpose of `notifyMany` was to handle the case where a consumer was subscribed to multiple properties on the same producer that changed at the same timestep.

I'm less convinced about the need for `notifyMany`, at least in a first iteration. I think the main win is where you have a Producer with a high number of reactive properties. Rather than needed to allocate observers for each property, you just need one observer that takes the changed property as an argument.

At least one issue with this approach is for something that is morally an `Observable<number>` - you need to access it through an additional property (say "value"). Though, I figured this case would be the less interesting/frequent one, and most scenarios would involve something that is map-like.

These were just my design ideas that I had so far. The core implementation of tiny-calc expressions is not dependent on this, so it would be fine to move away from this. It's a genuine open design question I have so I'm happy to replace it with something better suited.

jack-williams commented on packages/core/nano/src/types.ts:
```diff
@@ -1,22 +1,139 @@
+  /**
+   * Invoked whenever the data this object is bound to is changed.
+   */
+  notify: <K extends keyof T>(property: K, value: T[K]) => void;
```
If a consumer can subscribe to multiple producers then I think the paramater would be needed in notify. In this case (and perhaps regardless), I wonder if producers should have a string id on them to make them easy to identify (rather than relying on object identity).

jack-williams commented on packages/core/nano/src/types.ts:
```diff
@@ -1,22 +1,139 @@
+   * @param dataBindingReference - An instance of dataBindingReference, a reference
+   * to some data in the bound component.
+   */
+  read<K extends keyof T>(consumer: IConsumer<T>, property: K, ...args: any[]): T[K] | Pending<T[K]>;
```
I guess the semantics are that `read` subscribes only to the given property (merging with any existing subscriptions on the producer), and that `addConsumer` subscribes to everything?

DLehenbauer replied:
No, I just goofed.  I copied your signature for 'request()' without reconciling it with Bohemia's explicit subscription.

In my experience, it's almost always less expensive for the consumer to filter notifications than it is for the producer to track subscriptions to fine-grained properties, so I would recommend that contact always allows the producer the option of over-invalidating.

Of course, a producer can choose to be more frugal with its notifications, if it happens to be convenient.  We could also offer a generic IConsumer/IProducer that passes through the data from a source IProducer, but adds fine-grained dependency tracking to squelch notifications.

jack-williams replied:
Sounds good to me. So the contract of `read` is: get property `K` and subscribe to the producer. Subsequent notifications to the consumer will be bounded below by `K`, but possibly more.

DLehenbauer replied:
BTW - A cute trick for reducing the overhead of the implicit subscription model is to require the consumer to acquire a reader object from the producer, like this:
```ts
const reader = producer.open(consumer);
const length = reader.read('length');
for (let i = 0; i < length; i++) {
    a[i] = reader.read(i);
}
```
The producer adds the consumer to its set one time in 'open()' and then returns the same/cached reader object to every consumer.

You can deter consumers from caching readers by expiring them at the end of a recalc cycle, and mixing the allocation.

jack-williams replied:
Cool! In the implicit subscription model do we want per property invalidation in `notify`? Should it be the changed subset of all properties, or just the entire producer/reader and we let the consumer figure out what changed?

jack-williams replied:
And would the open/reader approach be part of the base interface?

DLehenbauer replied:
What's worked well for me in the past is to have the Producer call the consumer with each changed key

jack-williams replied:
So as it is now? Works for me

jack-williams commented on packages/core/nano/src/types.ts:
```diff
@@ -1,22 +1,139 @@
+   * @param dataBindingReference - An instance of dataBindingReference, a reference
+   * to some data in the bound component.
+   */
+  read<K extends keyof T>(consumer: IConsumer<T>, property: K, ...args: any[]): T[K] | Pending<T[K]>;
```
I think we can get rid of the `...args` at the end, unless you see a use case.

jack-williams commented:
Just added some minor changes:
- Removed the rest parameter: it's legacy from stuff I was trying out.
- Add `IProducer` to `notify` for multi-producer single consumer. TBD if this is what we would want, but I've changed the signature for now.

@dpgil @DLehenbauer I'm happy for you iterate and push any more changes.

jack-williams commented:
Question: What is the difference / overlap between a data binding service and a producer? To what extent are they similar?

dpgil commented:
In Bohemia context, we need some way to discover what producers available for binding. It's really just a list that Producers can subscribe to when they are created

dpgil commented:
Took my turn. 
-Removed onDeleted and onProducerDeleted
-Changed getValue to return IData | Pending<IData>
-Updated the exports in index.ts

At this point I agree that IDataResult is an app-specific implementation, so we may want to consider removing it.
Question: Do we want any data shapes in our generic types. i.e. do we want something like:
`export interface IProducer<T extends IData>`?

@DLehenbauer

jack-williams commented:
IMO I think the base type of Producer should make no assumptions about the data if possible.

A random thought is whether if property enumeration was added to producers, can we implement data-binding services and data results as specialised producers (the latter being a producer that tracks no dependencies). E.g.
```ts
// An empty implementation
declare const nullConsumer: IConsumer<DataResult>;

export class DataResult implements IProducer<Record<string, DataResult>> {

    // some underlying data store we are providing.
    constructor(private store: Record<string, DataResult>) { }

    // no-op DataResults don't track deps
    addConsumer() { }

    // no-op DataResults don't track deps
    removeConsumer() { }

    getProperties(): Promise<string[]> {
        // the keys in the store
        return Promise.resolve(["todo"]);
    }

    getChildren(): Promise<DataResult[]> {
        const getter = (p: string) => this.child(p)
        return this.getProperties().then(props => props.map(getter));
    }

    read(_consumer: IConsumer<DataResult>, prop: string): DataResult {
        return this.store[prop]; // or some other implementation
    }

    child(prop: string): DataResult {
        return this.read(nullConsumer, prop)
    }

    // implemented by some data passed in the constructor
    getValue() {}
    getHint() {}
    getFriendlyName() {}
}

declare const data: DataResult;
data.child("foo").child("bar").getValue();
```

One thing I wanted to support in tiny-calc is the ability to pass producers directly into formulas as their context so that formulas could be used to query the services available directly.

DLehenbauer commented:
My changes:
- Switched to implicit subscription via 'open()'
- Removed IDataResult and IData (I think we're agreed IDataResult belongs at a higher layer.)
- Took a first shot at 1D and 2D shapes.

I imagine there might be some more discussion on 1D and 2D, but at this point I'm pretty happy with the converged IProducer/IConsumer and am ready for Jack to have the final say.  How about you @dpgil?

jack-williams commented on packages/core/nano/src/types.ts:
```diff
@@ -1,22 +1,92 @@
+   * 
+   * @param consumer - The object to be notified of value changes.
+   */
+  open(consumer: IConsumer<T>): {
```
Is the invariant here that `prod.open(consumerA).read(p) === prod.open(consumerB).read(p)` if no updates have been sent? I guess it's likely to be the case that `prod.open(consumerA) === prod.open(consumerB)`, but I'm not sure we would need this as an invariant.

DLehenbauer replied:
The first invariant is effectively true.  If the two reads happen in separate turns, _technically_ one consumer could get a pending value while the second gets the final value, but both consumers will converge on the same value at the end of the recalc cycle.

Regarding if open() is required to return the same instance, I'd leave that up to the producer for now.

jack-williams replied:
Yep, that makes sense. I was just thinking about an even weaker invariant that only holds in the same turn. Basically a producer can't return a stateful reader, or return an observationally different reader to different consumers.

jack-williams commented:
I think this is looking good. It's easy enough to convert a producer with `open` to a calc object using a wrapper function so this isn't an issue.

I'm not sure whether the base `IProducer` type should come with a `getProperties` function. One reason for doing this would be so that _all_ producers are effectively a data-binding service - it just reduces the number of concepts someone needs to deal with. Though, it would be easy enough to extend `IProducer`, so I'm happy in either case.

I might suggest moving the `producer` argument in the `valueChanged` method to the end rather than the beginning. In the cases where we only subscribe to one producer, and therefore don't really care about the producer sending the update, we can write the `valueChanged` as `(property, value) => ...` rather than `(_producer, property, value) => ...`.

DLehenbauer commented:
If '.open()' is in any way cumbersome, you should undo it.  It's only purpose is to serve as a micro-optimization for implicit subscription.

My only feedback on 'getProperties()' is that if we include it on IProducer, we probably need to make the following guarantees:

1. It will (eventually) enumerate all properties that match the query using an async pull model.
2. It returns an exact match in the first batch of results.

Something like:
```ts
getProperties(query: string): IIterable<Promise<string[]>>
```
I defer to you if you think this is a reasonable requirement for all producers and how useful this is to the parser/compiler/calc engine.  (Seriously, I really don't know.  It's certainly reasonable for Table-Document to do this.)

Regarding the order of arguments in 'valueChanged', I hadn't given it thought.  I'd say take an educated guess, and don't be shy about imposing your preferences!

jack-williams commented:
> If '.open()' is in any way cumbersome, you should undo it. It's only purpose is to serve as a micro-optimization for implicit subscription.

I think it should be fine and like the ability of being able to put turn specific information on the reader without having to fetch it per-property.

> My only feedback on 'getProperties()' is that if we include it on IProducer, we probably need to make the following guarantees: ....

I agree it should be lazy. Given that I'm not currently using property enumeration, and I'm also unsure about what other meta-data we would need in the future, I'm inclined to leave it out of the base interface.

I don't think I have the best sense about where all these interfaces belong. I would be happy with what we have currently staying in the tiny-calc repo, but I'm equally happy for them to be moved elsewhere if thats better. I'm fine with taking a backseat on this if anyone has strong opinions.

dpgil commented:
The change looks good to me, assuming @jack-williams is happy with it!

dpgil commented:
We'll just have to delete `IDataResult` and `IData` from the index.ts

jack-williams commented:
- Moved IConsumer before IProducer to match the latter cases.
- Moved `producer` argument to the end in `IConsumer` callbacks.
- Changed the exported types.

I'm happy with this - if anyone wants to make any final changes feel free.

Once it's done I can fix up `examples/` so that they run with the new types.

dpgil commented:
Two last things on my end:

1. `removeVectorConsumer` should take type `VectorConsumer`, not `IConsumer`
2. Should the return value from `open` have an interface type? I would like to use this for chained scenarios, like `producer.open(consumer).read(p1).read(p2)...` since the property value are a tree for my case

jack-williams commented:
1. Good catch, I'll fix that up.
2. Yeah if it helps I think it's fine to add something like `IReader`.

DLehenbauer commented:
FYI - The chained read example probably needs a 2nd open:
```ts
producer
    .open(consumer).read(p1)
    .open(consumer).read(p2);    // Must also subscribe to the IProducer returned as 'p1'
```
But +1 on IReader.  The only motivation for defining the type inline was laziness.

jack-williams commented:
Updated the types and the examples so that they compile.

On the chained example I think there are two options to avoid multiple `open` calls. In the case where you only need notifications at the top-level you can define a root producer that just returns a tree of `IReader`'s. If you want nested Producers then you can recursive wrap the root producer to return something that implicitly calls `open`. I have an example like that [here].(https://github.com/microsoft/tiny-calc/blob/55353e9ad7e2835de60141d618896fe8abda3d40/packages/test/example/src/types.ts#L57)

dpgil commented:
Thanks Jack, looks good!

egnezami commented:
> FYI - The chained read example probably needs a 2nd open:
> 
> ```ts
> producer
>     .open(consumer).read(p1)
>     .open(consumer).read(p2);    // Must also subscribe to the IProducer returned as 'p1'
> ```
> 
> But +1 on IReader. The only motivation for defining the type inline was laziness.

@DLehenbauer 
I thought the chain components could be constructed like this:
```ts
producer.open(consumer)
    .read(p1)
    .read(p2);'

Is there a reason to call open every step in the chain?

jack-williams commented:
I think opening at each level enables different consumers to be bound at different levels of the tree. If we take a databinding service as a producer that returns, amongst other things, a table called "Foo" that is also a producer then we have something like:
```ts
service.open(consumerA).read("Foo").open(consumerB).read("colA");
```
Consumer A gets notified when the service changes, for instance when tables gets deleted or completely replaced. Consumer B only gets notified specifically when table "Foo" changes.

You can still have the structure:
```ts
producer.open(consumer)
    .read(p1)
    .read(p2);
```
if you want, but the cost comes at giving `consumer` a higher number of updates. One option would be to just wrap the root producer and implicitly use the same consumer all the way down. Another option, which I think is closer to what you have now, is to just have one producer at the top level, and have values that are not producers form the tree. Something like:
```ts
interface DataResult extends IReader<Record<string, DataResult>> {}
interface Root extends IProducer<Record<string, DataResult>> {}

declare const service: Root;
declare const consumer: IConsumer<Record<string, DataResult>>;
service.open(consumer).read("foo").read("bar").read("baz");
```

egnezami commented:
@jack-williams, what I had in mind was very close to the last option you provided.   thanks for the explanation.

egnezami commented:
Are there any other concerns with this? If not, @jack-williams, will you be able to merge this, or can we do it on our end?

jack-williams commented:
@egnezami I’m happy with this and can merge. I’m still reasonable flexible so if things need to change down the line because we discover something that’s fine from my POV.

Any remaining changes @DLehenbauer?

If everyone is happy I can merge. (Might be a 2hour delay as I’m just about the get on a flight, but DanielL also has admin if it’s urgent)

DLehenbauer commented:
I think it's good to go.

egnezami commented:
Thank you @jack-williams . It can wait a few hours.

jack-williams commented:
Done!
